File: spark-analytics/kafka-spark-processor/java-app/src/main/java/com/example/processor/model/PurchaseOrder.java
package com.example.processor.model;

/**
 * Represents a petrol pump purchase order.
 * Wire format (pipe-delimited): petrolPumpId|machineId|city|purchaseTime|fuelType|qnty|amt|pType
 */
public record PurchaseOrder(
    String petrolPumpId,
    String machineId,
    String city,
    String purchaseTime,
    int fuelType,
    int qnty,
    int amt,
    int pType
) {
    public static PurchaseOrder fromPipeString(String line) {
        String[] f = line.split("\\|");
        if (f.length < 8) throw new IllegalArgumentException("Invalid order: " + line);
        return new PurchaseOrder(f[0], f[1], f[2], f[3],
                Integer.parseInt(f[4].trim()),
                Integer.parseInt(f[5].trim()),
                Integer.parseInt(f[6].trim()),
                Integer.parseInt(f[7].trim()));
    }
}

File: spark-analytics/kafka-spark-processor/scala-app/src/main/scala/com/example/processor/model/PurchaseOrder.scala
package com.example.processor.model

/**
 * Petrol pump purchase order model.
 * Wire format: petrolPumpId|machineId|city|purchaseTime|fuelType|qnty|amt|pType
 */
case class PurchaseOrder(
  petrolPumpId: String,
  machineId: String,
  city: String,
  purchaseTime: String,
  fuelType: Int,
  qnty: Int,
  amt: Int,
  pType: Int
)

object PurchaseOrder {
  def fromPipeString(line: String): PurchaseOrder = {
    val f = line.split("\\|")
    require(f.length >= 8, s"Invalid order format: $line")
    PurchaseOrder(
      petrolPumpId = f(0), machineId = f(1), city = f(2), purchaseTime = f(3),
      fuelType = f(4).trim.toInt, qnty = f(5).trim.toInt,
      amt = f(6).trim.toInt, pType = f(7).trim.toInt
    )
  }
}
